### ISSUE TITLE
fieldName hydrates properly, but field-name doesn't.

### ISSUE THREAD

**Issue by bitwombat** [doctrine/DoctrineModule#601]
I noticed that if my form's field name was called "hire-date", that this line wasn't converting the POSTed string to a DateTime object:

https://github.com/doctrine/DoctrineModule/blob/master/src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php#L285

That's because $field from reflection of the entity is "hireDate".  They don't match.

But what makes me think that hire-date should work is that the very next line in the source determines the setter name using the Inflector::classify method, which is happy to turn "hire-date" into "hireDate".

Should the field be set using Inflector::classify before looking up its name with $metadata->getTypeOfField($field)?

**Comment by TomHAnderson** [comment]
I think that's a good idea.   Would you like to post up the change with some unit tests for your hyphen-ated field naming?

**Comment by bitwombat** [comment]
I'll try!

**Comment by TomHAnderson** [comment]
Please be aware that naming strategies keep popping up as problems.  I can't tell you what to do about it but as this is a naming strategy issue I wanted you to be aware.

**Comment by bitwombat** [comment]
@TomHAnderson I believe it to be a real bug so I worry about hiding the issue by closing it.  The failure is mine for not finding time yet to address the solution.

**Comment by TomHAnderson** [comment]
We're coming up on a 2.0 release.  There is a pending PR for using the inflector instead of ucfirst in the hydrator.  Would you try reproducing your issue with this PR?

https://github.com/doctrine/DoctrineModule/pull/625

**Comment by TomHAnderson** [comment]
A note on this issue:
```
<?php

$hire-date = 3;

echo $hire-date;
```

You can't have a field named $hire-date.   There is a way to make a field called $hire-date:

```
${"hire-date"} = 3;

echo ${"hire-date"};
```

I've become suspicious of your entity field name.  Can you clarify what the name of the field on your entity is?

**Comment by bitwombat** [comment]
> I've become suspicious of your entity field name. Can you clarify what the name of the field on your entity is?

On the entity it's hireDate, on the zend-form field it's hire-date.

Will make a testcase and test the PR.  Looks like I may have lost out on the opportunity to contribute my first (non-documentation) Doctrine PR!

**Comment by bitwombat** [comment]
I don't often use ZF and Doctrine programmatically, so I need a bit of help.

Just to mess around with DoctrineHydrator I need an EntityManager instance... which is a quite a beast requiring a connection etc.  What should I do there?

**Comment by bitwombat** [comment]
OK, got a simple testcase working.  Probably need to turn it into a PHPUnit test?

Digging deeper, the array accessed by this line:

https://github.com/doctrine/doctrine2/blob/2.6/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L2023

has a key of `hireDate`, whereas the field name (the key in $data, from zend-from) is `hire-date`
So, the type of it isn't determined properly.

**Comment by bitwombat** [comment]
> So, it sounds to me like you're missing a piece which is supposed to turn 'hire-date' into 'hireDate'.

Well, the Inflector does turn hire-date (or hire_date) to HireDate in the very next line for the purpose of finding the setter.  That's why it looked to me like the code meant to support hyphenated HTML field names.

It's just the call to the Inflector needs to be before the call to handleTypeConversions, that's all.

The other place I got this idea was from the zend-form documentation that uses hyphens in its multi-word field names (and [Google's HTML/CSS style guide](https://google.github.io/styleguide/htmlcssguide.html#ID_and_Class_Name_Delimiters) too).

Yes, using hireDate works but introduces a third style into my HTML & CSS (hyphens, underscores and now camel case).


> IMHO if you use canonical naming across your application the issue will not exist.

Using canonical (same) naming in different domains means violating conventions in those domains (or in pre-existing code).

Certainly not trying to mountain any molehills, but with a one-line change (in my experiments) the HTML/CSS can have a self-consistent style and Doctrine is happy too.

More interested in making friends and keeping an awesome dev/maintainer happy than winning arguments!

**Comment by TomHAnderson** [comment]
Cool.  So what I need from you is a way to test your issue.  Even if that means you put your whole application up on github with a button to duplicate it, or just a unit test, that's fine.  For me to understand the issue, because I no longer use zend-form or the form support in this module, walking through your example will clear up the problem for me.

Thanks!

**Comment by TomHAnderson** [comment]
Or....you said it's a matter of swapping two lines.  I think I get what you're saying and if it's a bug that the lines are in the wrong order maybe we can skip all the test (however a unit test to show it works would still be good).

> It's just the call to the Inflector needs to be before the call to handleTypeConversions, that's all.

Please give me a link to the line number.

**Comment by bitwombat** [comment]
> Please give me a link to the line numbers.

Same as the original comment:

https://github.com/doctrine/DoctrineModule/blob/master/src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php#L285

**Comment by bitwombat** [comment]
Hi Tom - here's a fork with my guess at two new tests and my guess at a fix.  The second one fails without the fix to DoctrineObject.  

I can make it a PR but I didn't want to pollute.  

I'm pretty unsure of what I've done - names, logic, approach, etc. - so all criticism is welcome.

https://github.com/bitwombat/DoctrineModule/tree/hydrate_hyphens

**Comment by TomHAnderson** [comment]
Your fix includes mutating the `$field.  Though it's true we mutate the `$value` I'm worried this change is a bit too specific to put a change in which would reword any field.  For instance a field called `FieldName` would be changed to `fieldName`.

The solution I'm going to recommend is you normalize your posted values before sending them to the hydrator.  Here is a snippet to do just that:

```
include __DIR__ . '/vendor/doctrine/inflector/lib/Doctrine/Common/Inflector/Inflector.php';
include __DIR__ . '/vendor/doctrine/common/lib/Doctrine/Common/Util/Inflector.php';

use Doctrine\Common\Util\Inflector;

$data = [
    'fieldOne' => 1,
    'field-two' => 2,
    'field-three-eee' => 3,
];

$result = [];
array_walk($data, function($value, $key) use (&$result) {
    $key = lcfirst(Inflector::classify($key));
    $result[$key] = $value;
});

print_r($result);
```
